dashboard/test/controllers/ai_diff_controller_test.rb

```diff
@@ -0,0 +1,158 @@
+require 'test_helper'
+
+class AiDiffControllerTest < ActionController::TestCase
+  include Devise::Test::ControllerHelpers
+
+  setup do
+    @course_offering = create(:course_offering, display_name: 'Course Name')
+    @course_version = create(:course_version, :with_unit_group, course_offering: @course_offering)
+    @unit_group = @course_version.content_root
+    @unit_in_course = create(:script, name: 'unit-in-teacher-instructed-course2')
+    create(:unit_group_unit, script: @unit_in_course, unit_group: @unit_group, position: 1)
+    @lesson_group = create(:lesson_group, script: @unit_in_course)
+    @lesson = create(:lesson, script: @unit_in_course, lesson_group: @lesson_group)
+    create(:script_level, script: @unit_in_course, lesson: @lesson)
+
+    @teacher = create(:teacher)
+    @unit_display_name = "Beowulf Course"
+
+    @session_id = "1234"
+    @bedrock_client = Aws::BedrockAgentRuntime::Client.new(stub_responses: true)
+    @bedrock_client.stub_responses(
+      :retrieve_and_generate, {
+        citations: [
+          {
+            generated_response_part: {
+              text_response_part: {
+                span: {
+                  end: 55,
+                  start: 0
+                },
+                text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit"
+              }
+            },
+            retrieved_references: [
+              {
+                content: {
+                  text: "Hwaet! We gar-dena in geardagum, theod-cyninga thrym gefrunon"
+                },
+                location: {
+                  s3_location: {
+                    uri: "s3://dummy_file"
+                  },
+                  type: "S3"
+                }
+              }
+            ]
+          }
+        ],
+        output: {
+          text: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua."
+        },
+        session_id: @session_id
+      }
+    )
+    AiDiffBedrockHelper.stubs(:create_bedrock_client).returns(@bedrock_client)
+  end
+
+  test "returns bad_request when getting chat_completion if bad params" do
+    sign_in @teacher
+
+    SingleUserExperiment.expects(:enabled?).with(user: @teacher, experiment_name: 'ai-differentiation').returns(true)
```

Comment:
once you start using CanCan, you won't be able to stub this because ability.rb checks for other experiments. instead, you can just create the necessary experiment instead:
```suggestion
    SingleUserExperiment.create!(...)
```
experiments get reset between test cases, along with everything else in the DB.

Reply:
I ended up doing this with the factories in setup, which seems to be ok